src/tactic/sort.lean

```diff
@@ -0,0 +1,179 @@
+last in the sum.
+
+We use this function for expressions in an additive commutative semigroup. -/
+meta def sorted_sum_with_rel
+  (hyp : option name) (rel : expr → expr → bool) (t : list expr) (e : expr) : tactic unit :=
+match (get_summands e).sort_with_top t rel with
+| eₕ::es := do
+  e' ← es.mfoldl (λ eₗ eᵣ, mk_app `has_add.add [eₗ, eᵣ]) eₕ,
+  e_eq ← mk_app `eq [e, e'],
+  n ← get_unused_name,
+  assert n e_eq,
+  e_eq_fmt ← pp e_eq,
+  reflexivity <|>
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
+      fail format!"Failed to prove:\n {e_eq_fmt}",
+  h ← get_local n,
+  match hyp with
+  | some loc := do
+    get_local loc >>= rewrite_hyp h,
+    tactic.clear h
+  | none     := do
+    rewrite_target h,
+    tactic.clear h
+  end
+| [] := skip
+end
+
+/-- Partially traverses an expression in search for a sum of terms.
+When `recurse_on_expr` finds a sum, it sorts it using `sorted_sum_with_rel`. -/
+meta def recurse_on_expr (t : list expr) (hyp : option name) (rel : expr → expr → bool) :
+  expr → tactic unit
+| e@`(%%_ + %%_)     := sorted_sum_with_rel hyp rel t e
+| (expr.lam _ _ _ e) := recurse_on_expr e
+| (expr.pi  _ _ _ e) := recurse_on_expr e
+| e                  := e.get_app_args.mmap' recurse_on_expr
+
+/-- Calls `recurse_on_expr` with the right expression, depending on the tactic location. -/
+meta def sort_summands_aux (rel : expr → expr → bool) (t : list expr) : option name → tactic unit
+| (some hyp) := get_local hyp >>= infer_type >>= recurse_on_expr t hyp rel
+| none       := target >>= recurse_on_expr t none rel
+
+end with_rel
+
+/--  The order on `polynomial.monomial n r`, where monomials are compared by their "exponent" `n`.
+If the expression is not a monomial, then the weight is `⊥`. -/
+meta def monomial_weight (e : expr) : option ℕ :=
+match e.app_fn with
+| `(coe_fn $ polynomial.monomial %%n) := n.to_nat
+| _ := none
+end
+
+/--  The function we use to compare two `expr`:
+* all non-monomials are compared alphabetically;
+* all non-monomials are smaller than all monomials;
+* bigger monomials have higher exponent.
+-/
+meta def compare_fn (eₗ eᵣ : expr) : bool :=
+match (monomial_weight eₗ, monomial_weight eᵣ) with
+| (none, none)     := eₗ.to_string ≤ eᵣ.to_string -- this solution forces an unique ordering
+| (_, none)        := false
+| (none, _)        := true
+| (some l, some r) := l ≤ r
+end
+
+/--  A version of `sort_summands_aux` that allows failure, if `allow_failure = tt`. -/
+meta def sort_summands_core (allow_failure : bool) (t : list expr) (hyp : option name) :
+  tactic unit :=
+if allow_failure then sort_summands_aux compare_fn t hyp <|> skip
+else sort_summands_aux compare_fn t hyp
+
+/-- If the target is an equality involving summands,
+then  `sort_summands` sorts the summands on either side of the equality.
```

Comment:
It's no longer restricted to equalities. `recurse_on_expr` should make it work with many shapes (=, ≠, <, ≤ etc)

Reply:
Amended!